### ISSUE TITLE
[Enhancement] Please add integration with GeoClue location service.

### ISSUE THREAD

**Issue by Baggypants** [navit-gps/navit#1032]
GeoClue allows applications to integrate with multiple location sources, such as Mozilla Location Service, 3G, GeoIP and various GPS systems. Crucially for me it is being provided as the main location source in the Pinephone Mobian distribution and I'd like to use Navit for it. 

GeoClue https://gitlab.freedesktop.org/geoclue/geoclue/-/wikis/home

Mobian Location docmentation for Pinephone https://wiki.mobian-project.org/doku.php?id=location

Pinephone Navit discussion https://forum.pine64.org/showthread.php?tid=10697&pid=74115#pid74115

**Comment by mvglasow** [comment]
Just for the background, we are talking primarily about the Linux build, correct? Then GeoClue would essentially replace gpsd or whatever other location source we support on Linux.

Sounds like we are talking about another `vehicle` implementation (Navit’s way of encapsulating the various location sources found in the wild). That should be sorted out on the architecture side already.

What platforms would we need to support with this?

**Comment by Baggypants** [comment]
Yes, it would be targeting Linux and would be an alternative source to gpsd or reading nmea data out of a serial device on that platform. gpsd and GeoClue have different use cases though. gpsd allows location data to be collected and transmitted across a network. GeoClue aggregates different sources of location data on a local machine to one interface.

**Comment by jkoan** [comment]
I have already looked into the GeoClue API, and from what i see i think it should be quite strait forword. As @mvglasow said this would be another vehicle plugin for Navit. 
My only concern so far is that we don't have the real hardware (Pinephone or some other device which has a image with GeoClue). I ordered a PineTab from the first batch, so i could probably sort some things out, but the PineTab doesn't have GNSS as the PinePhone has.
Do you know if there is any OS to test/prototype with, or some sort of fake Data feedin? Probably we could feed a pipe into GeoClue and feed the pipes input from previous collected nema data. But this approche would take at least some time setting up all this, and is probably a bit sluggish.

**Comment by metalstrolch** [comment]
GeoClue is used for Qt location services on SailfishOs. Don't know if it is v1 or v2 unfortunately. Since the development headers for GeoClue should be available, one might test a GeoClue vehicle plug in on Sailfish OS devices. It's no "win" for SailfishOS, as the Qt5 vehicle is already there and working, but it's OK to give it a try. Sailfish OS is "just Linux" in all cases.
When back from vacation, I'm willing to give this a try. If somebody already started such a plug in, then drop me a note, so I might set up testing on my sailfish device.

**Comment by sirnewton01** [comment]
Someone has written a GeoClue mock from long ago. https://github.com/inzanity/geoclue-mock

Also, GeoClue will ultimately fall-back to using available WiFi or IP address for your location (inaccurate as they may be) without a GNSS modem on your system. I do this from time to time on my desktop.

Maybe these would be enough to test out a new navit geoclue vehicle?

**Comment by jkoan** [comment]
I have already started to write the vehicle plugin after i realized that my gnome desktop has it preconfigured (Manjaro Linux). I already have the main things set up. Only need to feed the inputs from GeoClue to navit.
Hopefully i can provite a prototype soon.

**Comment by jkoan** [comment]
First Prototype is done and working, will take some more hours tomorrow to tidy everything up, and then i will provide a Pull Request for testing. 
My first test was on a local LAN PC where i switched between USB tethering (mobile phone) and LAN/DSL connection. This caused GeoClue to update the position and Navit did accordingly :smile: 
Tomorrow i will clean everything up, and test on my Laptop because it has a GNSS/GPS module build in, so i should be able to do some movement :wink:

**Comment by Baggypants** [comment]
Is the prototype code available in a repository?

**Comment by jkoan** [comment]
Got the Geoclue Prototype up at the branch https://github.com/navit-gps/navit/tree/vehicle_geoclue

**Comment by metalstrolch** [comment]
Bummer. SailfsihOS is of no help here unfortunately. It still uses ages old geoclue version 0.12.99. Don't think we want to support that old versions. Even more as Sailfish works wit qt5 vehicle.
Stated branch requires at least geoclue-2.

**Comment by sirnewton01** [comment]
@jkoan I think maybe the speed is getting assigned to the altitude in error here? https://github.com/navit-gps/navit/compare/vehicle_geoclue#diff-ad9d0ae6bf9e20aae39c3f2d8d135c8cR89

**Comment by jkoan** [comment]
> @jkoan I think maybe the speed is getting assigned to the altitude in error here? https://github.com/navit-gps/navit/compare/vehicle_geoclue#diff-ad9d0ae6bf9e20aae39c3f2d8d135c8cR89

Yes indeed, this ist an Error because i was only able to test stationary so far

**Comment by jkoan** [comment]
I pushed a update to the branch with the fix :)

**Comment by sirnewton01** [comment]
Thanks, sorry, it stood out when I was looking at the changes.

I don't really know anything about this code base. How does navit decide to use a vehicle like geoclue instead of gpsd? Is it decided at installation or configuration time?

**Comment by jkoan** [comment]
> Is it decided at installation or configuration time?

Ist decided at configuration time with the vehicle source
https://wiki.navit-project.org/index.php/Configuration/Vehicle_Options#Vehicle_Options

The source is just called "geoclue://"

**Comment by jkoan** [comment]
@Baggypants @sirnewton01 any updates how well its working?

**Comment by sirnewton01** [comment]
@jkoan I'm trying to get the new branch compiling but I'm running into some troubles. Latest one is this

```multiple definition of `map_information_attrs```

**Comment by jkoan** [comment]
You don't need maptool, try to disable it for now, #1045 is already open for this issue.
`cmake -DBUILD_MAPTOOL=0`

**Comment by sirnewton01** [comment]
Now it says this.

```
[  7%] Generating airport.png
Unknown option --without-gui
```

Mabye I should just get the Debian source package and try it from there with your patches.

**Comment by sirnewton01** [comment]
I've got an experimental Debian package with the patch from the branch. Next step is to put together the configuration and give it a try.

**Comment by jkoan** [comment]
> Now it says this.
> 
> ```
> [  7%] Generating airport.png
> Unknown option --without-gui
> ```

This seems to be the conversation from svg to png which dynamically uses whatever is available on your system. 
Could you do `cmake -L | grep IMAGE` on the failed build dir and provide the output?

**Comment by sirnewton01** [comment]
> Could you do `cmake -L | grep IMAGE` on the failed build dir and provide the output?

```
IMAGE_CONVERTER:FILEPATH=/usr/bin/convert
IMAGE_CONVERTER_SVGZ:FILEPATH=/usr/bin/inkscape
IMAGE_DIR:PATH=share/navit/icons
```

Somehow the debian package maintainers worked around this problem, but I haven't looked into it.

**Comment by sirnewton01** [comment]
Either I patched navit wrong somehow when enabling the package or I did the configuration wrong. Here's my vehicle line in my navit.xml:

```<vehicle name="Local GPS" profilename="car" enabled="yes" active="1" source="geoclue://">
```

Here's the error log from navit when I launch it:

```
Gtk-Message: 15:51:12.546: Failed to load module "canberra-gtk-module"
error:navit:traffic_new:wrong type 'null'
error:navit:vehicle_new:invalid source 'geoclue://': unknown type 'geoclue'
error:navit:xinclude:Unable to include '/usr/share/navit/maps/*.xml'
error:navit:attr_new_from_text:Unknown item type 'poly_swimming_pool' ignored.
```

**Comment by sirnewton01** [comment]
Ok, now I think I have the geoclue vehicle included. The latest error is this:

```
error:navit:traffic_new:wrong type 'null'
error:vehicle_geoclue:plugin_init:enter
error:navit:xinclude:Unable to include '/usr/share/navit/maps/*.xml'
error:navit:attr_new_from_text:Unknown item type 'poly_swimming_pool' ignored.
error:navit:xml_parse_text:FATAL: No XML data supplied.
```

**Comment by sirnewton01** [comment]
Ok, my GUI layout is all messed up, but I loaded in some OSM maps and set a center coordinates that aren't precisely where I am. I'm inside my house and I know that it can't have possibly gotten a GPS fix, but geoclue is able to do a position based on my nearby wifi networks. It works! I can tell by nearby landmarks and the local towns that the position is right.

Thank you.

Apparently I have alot to learn about configuring Navit though. :)

**Comment by jkoan** [comment]
> I had a go at setting up an aarch64 vm and compiling navit for the pinephone but It doesn't quite work. If I try and set the `vehicle` source to geoclue:// then it loads, displays the map and crashes. If I set the source to the ttyUSB1 then it doesn't.
> 
> ```shell
> (navit:2547): Gtk-WARNING **: 21:27:33.441: Unable to locate theme engine in module_path: "adwaita",
> Gtk-Message: 21:27:33.444: Failed to load module "canberra-gtk-module"
> error:navit:command_register_callbacks:could not resolve object in cs=0xaaaabd6d7f50:
> error:navit:command_register_callbacks:follow>1
> error:navit:command_register_callbacks:^
> error:vehicle_geoclue:plugin_init:enter
> error:navit:xml_parse_file:could not open XML file
> error:navit:traffic_get_messages_from_xml_file:could not retrieve stored traffic messages
> error:navit:osd_std_config:Error in saved command: not_ready, item=0xaaaabd6d7e40.
> error:navit:osd_std_config:Error in saved command: not_ready, item=0xaaaabd6d7e40.
> error:navit:graphics_image_new_scaled_rotated:No image for '/home/mobian/navit/icons/'
> sh: 1: gdb: not found
> calling gdb returned 32512
> ```
> 
> It is trying to do something with geoclue though because if I turn off the wifi it doesn't crash.

I was not able to reproduce this on my x86 PC running manjaro with geoclue 2.5.6

**Comment by Baggypants** [comment]
I wasn't able to reproduce it on my x86_64 either, as I [noted](https://github.com/navit-gps/navit/issues/1032#issuecomment-678526140)

**Comment by jkoan** [comment]
Compiled on the Pinetab (no GPS only WIFI based position) , and got it working so far. Got a initial position, still need to do a walk with the tab. Will write again once done.

**Comment by jkoan** [comment]
I just did my tests and had no crashes. Perhaps you config was broken?
I attached my Config just here: [navit.xml.zip](https://github.com/navit-gps/navit/files/5428846/navit.xml.zip)

I compiled based on 55d64dd5476a9f8cca0db1c93f345ae1defccab3 (current trunk) where i rebased the branch vehicle_geoclue

**Comment by jkoan** [comment]
@Baggypants could you try this build as well? It includes the same config i posted in the last Comment
[navit.zip](https://github.com/navit-gps/navit/files/5429177/navit.zip)
Map location is configured as /home/mobian/.navit/*.bin

**Comment by Baggypants** [comment]
This build works really well on the pinephone. It launched and picked up a lock instantly within 4 miles and then within 30m in under two minutes.

**Comment by sirnewton01** [comment]
If you are using mobian then you'll have some modemmanager enhancements for LTE cell phone tower location via MLS. Otherwise, it might take some time for mm to make a release with the patch.

**Comment by jkoan** [comment]
> This build works really well on the pinephone. It launched and picked up a lock instantly within 4 miles and then within 30m in under two minutes.

Nice, thx much for testing. Have you tried navigation as well? And it so, is it working flawlessly?

**Comment by Baggypants** [comment]
I had a go but it all goes to sleep too quickly, I'm going to try running it with caffeine on another day.

**Comment by Baggypants** [comment]
I kept it awake with `gnome-session-inhibit --inhibit-only` and not pressing the button to lock the screen. Navit with geoclue was able to pinpoint my position in my house so I could program in a route and have it calculate, and then once I took it outside, it was able to get a gps lock (eventually) to do some turn-by-turn navigation. I don't have the mobile data network set up on the phone yet so it was reliant only on gps outside. I guess if I did have mobile data it would have been able to recognise nearby AP's and towers which would have made the transition easier. Battery lasted roughly 3 hours.
